StatefulSet: Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings StatefulSet into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

Fixes #36859

**Special notes for your reviewer**:

**Release note**:

```release-note
StatefulSet now respects ControllerRef to avoid fighting over Pods. At the time of upgrade, **you must not have StatefulSets with selectors that overlap** with any other controllers (such as ReplicaSets), or else [ownership of Pods may change](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#upgrading).
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
StatefulSet: Don't touch Pods if DeletionTimestamp is set.

Commit:
StatefulSet: Add ControllerRef on all created Pods.

Commit:
StatefulSet: Set DefaultGarbageCollectionPolicy to OrphanDependents.

Now that StatefulSet adds ControllerRef to Pods it creates,
we need to set this default so legacy behavior is maintained.

Commit:
ControllerRefManager: Allow matching by more than just selector.

Commit:
StatefulSet: Use ControllerRefManager to adopt/orphan.

Commit:
StatefulSet: Use ControllerRef to route watch events.

This is part of the completion of ControllerRef, as described here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#watches

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42080)
<!-- Reviewable:end -->

Commit:
StatefulSet: Update Lister documentation for ControllerRef.

The StatefulSet Listers still use selectors, because this is the
behavior expected by callers. This clarifies the meaning of the
returned list. Some callers may need to switch to using
GetControllerOf() instead, but that is a separate, case-by-case issue.

kow3ns commented:
Reviewed 11 of 11 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/controller/statefulset/stateful_set.go, line 271 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080#-Kdm1kDEIOuCdOEctTHX:-Kdm1kDEIOuCdOEctTHY:b-crz5y9) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb95066be044617287f6bb45d56e77ba4e776e5e/pkg/controller/statefulset/stateful_set.go#L271)):*
> ```Go
> 	}
> 
> 	filter := func(pod *v1.Pod) bool {
> ```

you should use
```go
filter := func(pod *v1.Pod) bool {
        return isMember(set, pod)
}
```

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080)*
<!-- Sent from Reviewable.io -->

smarterclayton commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -163,15 +180,41 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		set, err := ssc.setLister.StatefulSets(curPod.Namespace).Get(curControllerRef.Name)
+		if err != nil {
```
Technically you should be handling NotFound here, but I don't think there's a downside.

kargakis replied:
Yeah, we usually just handle the `err == nil` case

enisoc replied:
@smarterclayton @kargakis Can you elaborate? I'm not sure if you two are agreeing with each other or not.

Are you saying I should skip the call to `HandleError()` if `IsNotFound(err)` is true?

kargakis replied:
Oh, I misread @smarterclayton. Usually in resource handlers we don't handle errors at all.

smarterclayton replied:
In general, any place where we utilruntime.HandleError, we should ask "if the item is not found, can we just return without logging anything safely instead".  In some places that is obvious (resource was deleted, we don't need to do anything further), in others it is not (we have to handle deletion safely).

enisoc replied:
In my case, I believe it's safe to do nothing if the error is NotFound. I was only intending to log transient errors, but @kargakis has reminded me elsewhere that this interface is backed by a cache that cannot fail for any reason other than NotFound. So, I plan to remove the call to HandleError in these cases.

Thanks for clarifying.

smarterclayton commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -163,15 +180,41 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
+	if curControllerRef != nil {
+		set, err := ssc.setLister.StatefulSets(curPod.Namespace).Get(curControllerRef.Name)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get StatefulSet %q in namespace %q: %v", curControllerRef.Name, curPod.Namespace, err))
```
style: All error messages should start lowercase (`couldn't get StatefulSet ...`)

smarterclayton replied:
Applies to all commits in this PR

enisoc replied:
Done.

enisoc commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/controller/statefulset/stateful_set.go, line 271 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080#-Kdm1kDEIOuCdOEctTHX:-KdmQetB38VyPkDpek1q:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eb95066be044617287f6bb45d56e77ba4e776e5e/pkg/controller/statefulset/stateful_set.go#L271)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

you should use
```go
filter := func(pod *v1.Pod) bool {
        return isMember(set, pod)
}
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -49,13 +49,18 @@ const (
 )
 
+// ControllerKind contains the schema.GroupVersionKind for this controller type.
+var ControllerKind = apps.SchemeGroupVersion.WithKind("StatefulSet")
```
unexport if you don't need it outside of the package

kargakis commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -142,15 +148,40 @@ func (ssc *StatefulSetController) Run(workers int, stopCh <-chan struct{}) {
+		set, err := ssc.setLister.StatefulSets(pod.Namespace).Get(controllerRef.Name)
+		if err != nil {
+			// Don't log NotFound. That just means the GC will soon delete the Pod.
+			if !errors.IsNotFound(err) {
```
The cache returns only 404s:
https://github.com/kubernetes/kubernetes/blob/77733c2afdcc4e85595e990d69d9b2765b1cf4e1/pkg/client/listers/apps/v1beta1/statefulset.go#L92
https://github.com/kubernetes/kubernetes/blob/77733c2afdcc4e85595e990d69d9b2765b1cf4e1/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L392

kow3ns commented:
Reviewed 4 of 6 files at r3.
Review status: 13 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 2 of 6 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080)*
<!-- Sent from Reviewable.io -->

Commit:
StatefulSet: Don't log Pod events unless some StatefulSet cares.

Commit:
StatefulSet: Always set BlockOwnerDeletion in ControllerRef.

enisoc commented:
@k8s-bot bazel test this
@k8s-bot node e2e test this
@k8s-bot kops aws e2e test this

kargakis commented on pkg/controller/statefulset/stateful_set.go:
```diff
@@ -153,16 +159,42 @@ func (ssc *StatefulSetController) Run(workers int, stopCh <-chan struct{}) {
+	if pod.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a new pod shows up in a state that
+		// is already pending deletion. Prevent the pod from being a creation observation.
+		ssc.deletePod(pod)
```
We should only care for pods owned by stateful sets

enisoc replied:
That check is done inside `deletePod()`. Is there a benefit to doing it before as well?

kargakis replied:
Nvmd, it's fine.

kargakis commented on pkg/controller/statefulset/stateful_set_control.go:
```diff
@@ -111,6 +112,12 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
+	// If the StatefulSet is being deleted, don't do anything other than updating
+	// status.
+	if set.DeletionTimestamp != nil {
+		return nil
```
@kow3ns we should have a helper that would do just that - iterate over all pods and update the status accordingly.

kargakis replied:
For now, it seems that this check is in the right place.

kow3ns commented:
/lgtm

---

Reviewed 4 of 7 files at r4, 2 of 2 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42080#-:-KeKi-ONsvSw1yGVKC8B:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

deads2k commented:
/approve

enisoc commented:
@k8s-bot node e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot cvm gce e2e test this

Commit:
StatefulSet: Use synchronous Delete of SS in e2e.

This is needed because we changed the default for SS to
OrphanDependents.

enisoc commented:
@k8s-bot kops aws e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot gce etcd3 e2e test this

Commit:
StatefulSet: Check that ControllerRef UID matches.

enisoc commented:
@kow3ns This just needs re-LGTM after some e2e fixes (see last couple commits) and rebases.

kow3ns commented:
/lgtm
